src/irmin-pack/pack_store.ml

```diff
@@ -136,83 +165,118 @@ module Maker
+          off len;
+      let key_of_offset off =
+        let hash = io_read_and_decode_hash ~off t in
+        Pack_key.v ~hash ~offset:off ~length:len
+      in
       let dict = Dict.find t.pack.dict in
-      Val.decode_bin ~hash ~dict (Bytes.unsafe_to_string buf) 0
+      Val.decode_bin ~key_of_offset ~dict (Bytes.unsafe_to_string buf) 0
 
     let pp_io ppf t =
       let name = Filename.basename (Filename.dirname (IO.name t.pack.block)) in
       let mode = if t.readonly then ":RO" else "" in
       Fmt.pf ppf "%s%s" name mode
 
-    let unsafe_find ~check_integrity t k =
-      Log.debug (fun l -> l "[pack:%a] find %a" pp_io t pp_hash k);
+    let unsafe_find ~check_integrity:_ t k =
+      Log.debug (fun l -> l "[pack:%a] find %a" pp_io t pp_key k);
       Stats.incr_finds ();
-      match Tbl.find t.staging k with
+      let hash = Key.to_hash k in
+      match Tbl.find t.staging hash with
       | v ->
-          Lru.add t.lru k v;
+          Lru.add t.lru hash v;
           Some v
       | exception Not_found -> (
-          match Lru.find t.lru k with
+          match Lru.find t.lru hash with
           | v -> Some v
           | exception Not_found -> (
               Stats.incr_cache_misses ();
-              match Index.find t.pack.index k with
-              | None -> None
-              | Some (off, len, _) ->
-                  let v = snd (io_read_and_decode ~off ~len t) in
-                  (if check_integrity then
-                   check_key k v |> function
-                   | Ok () -> ()
-                   | Error (expected, got) ->
-                       Fmt.failwith "corrupted value: got %a, expecting %a."
-                         pp_hash got pp_hash expected);
-                  Lru.add t.lru k v;
-                  Some v))
+              (* XXX *)
+              (* match Index.find t.pack.index k with
+               * | None -> None
+               * | Some (off, len, _) -> *)
+              let off, len = Key.(to_offset k, to_length k) in
+              if off >= IO.offset t.pack.block then
+                (* This key is from a different store instance, referencing an
+                   offset that is not yet visible to this one. It's possible
+                   that the key _is_ a valid pointer into the same store, but
+                   the offset just hasn't been flushed to disk yet, so we return
+                   [None]. *)
+                None
+              else
+                let v = snd (io_read_and_decode ~off ~len t) in
+                Lru.add t.lru hash v;
+                match equal_hash (Key.to_hash k) (Val.hash v) with
+                | false -> (* A clear happened *) None
+                | true -> Some v
+              (* XXX: this isn't a valid check, because it's possible that we
+                 read and decode a value with a different hash just because a
+                 clear happened and then there were writes over the offset. *)
```

Comment:
Perhaps we should, but it would mean changing the existing tests of `clear` & it might make keys hard to reason about from with Irmin core. We could also keep the generation number in the (in-memory representation of) the key, which distinguishes these issues but makes keys 1 word longer. We could also let the user pick whether they want `clear`, and only store the generation number if they do.

`clear` was added to support the layered store IIRC (and isn't yet used elsewhere), so perhaps @icristescu has an opinion on this.

Reply:
Resolved by dropping support for dereferencing keys after `Pack.clear`, and `Repo.clear` raises an exception to ensure the user doesn't try.